docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,408 @@
+    // Otherwise it selects all pods in the namespaces selected by namespaceSelector.
+    // +optional
+    NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty" protobuf:"bytes,2,opt,name=namespaceSelector"`
+
+    // ipBlock defines policy on a particular IPBlock. If this field is set then
+    // neither of the other fields can be.
+    // +optional
+    IPBlock *networkingv1.IPBlock `json:"ipBlock,omitempty" protobuf:"bytes,3,rep,name=ipBlock"`
+}
+
+// Status defines the observed state of NetworkQoS
+type Status struct {
+    // A concise indication of whether the EgressNetworkQoS resource is applied with success.
+    // +optional
+    Status string `json:"status,omitempty"`
+
+    // An array of condition objects indicating details about status of EgressNetworkQoS object.
+    // +optional
+    // +patchMergeKey=type
+    // +patchStrategy=merge
+    // +listType=map
+    // +listMapKey=type
+    Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+```
+
+* Ensure backward compatibility support available for `EgressQoS` API.
+
+### Implementation Details
+
+The new controller is introduced in OVN-Kubernetes which would watch `NetworkQoS` in addition to `EgressQoS`, `Pod` and `Node` objects, which will create the relevant NetworkQoS objects and attach them to all of the node local switches in the cluster in OVN - resulting in the necessary flows to be programmed in OVS.
+
+In order to not create an OVN NetworkQoS object per pod in the namespace, the controller will also manage AddressSets.
+For each QoS rule specified in a given `NetworkQoS` it'll create an AddressSet, adding only the pods whose label matches the PodSelector to it, making sure that new/updated/deleted matching pods are also added/updated/deleted accordingly. Rules that do not have a PodSelector will leverage the namespace's AddressSet.
+
+Similarly when `NetworkQoS` is created for Pods secondary network, OVNK must create a new AddressSet for every QoS rule. When no pod selector is specified, then it must contain all of the pod's IP addresses that belong to the namespace and selected network. If only a set of pods are chosen via podSelector, then it must have IP addresses only for chosen pod(s).
+
+For example, using LGW mode and assuming there's a single node `node1` and the following `NetworkQoS` is created:
+
+```yaml
+kind: NetworkQoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: qos-external-foo
+  namespace: default
+spec:
+  podSelector:
+    matchLabels:
+      priority: Critical
+  egress:
+  - priority: 100
+    dscp: 59
+    classifier:
+      ports:
+        protocol: "tcp"
+        port: 22
+      to:
+      - cidr: 1.2.3.4/32
+  - priority: 10
+    dscp: 46
+    classifier:
+      to:
+      - cidr: 1.2.3.4/32
+---
+kind: NetworkQoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: qos-external-bar
+  namespace: default
+spec:
+  egress:
+  - dscp: 30
```

Comment:
Please add `priority` in this example, since it is required.

Reply:
when `priority` is not specified, it's considered as `0`, isn't that ok ?

Reply:
It may be that the crd I'm using is not written properly, but as of now, the priority is needed:

**The NetworkQoS "default" is invalid: spec.egress[0].priority: Required value**


```
❯ cat n3.yaml
apiVersion: k8s.ovn.org/v1
kind: NetworkQoS
metadata:
  name: default
  namespace: default
spec:
  egress:
  - priority: 100
    dscp: 48
    classifier:
      to:
      - ipBlock:
          cidr: 2001:0db8:85a3:0000:0000:8a2e:0370:7330/124

❯ diff n3.yaml n3_no_priority.yaml
8,9c8
<   - priority: 100
<     dscp: 48
---
>   - dscp: 48

❯ k create -f ./n3_no_priority.yaml
The NetworkQoS "default" is invalid: spec.egress[0].priority: Required value

❯ k create -f ./n3.yaml
networkqos.k8s.ovn.org/default created
```

Reply:
ok, this priority value is going to be used while creating QoS entry in OVN, so making it as a required field.